readme.md
# Homebridge-automower

[![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)

[![npm](https://img.shields.io/npm/v/homebridge-automower.svg)](https://www.npmjs.com/package/homebridge-automower)
[![npm](https://img.shields.io/npm/dw/homebridge-automower.svg)](https://www.npmjs.com/package/homebridge-automower)
[![npm](https://img.shields.io/npm/dt/homebridge-automower.svg)](https://www.npmjs.com/package/homebridge-automower)

[![CodeFactor](https://www.codefactor.io/repository/github/nicoduj/homebridge-automower/badge)](https://www.codefactor.io/repository/github/nicoduj/homebridge-automower)
[![Build Status](https://travis-ci.com/nicoduj/homebridge-automower.svg?branch=master)](https://travis-ci.com/nicoduj/homebridge-automower)
[![Known Vulnerabilities](https://snyk.io/test/github/nicoduj/homebridge-automower/badge.svg?targetFile=package.json)](https://snyk.io/test/github/nicoduj/homebridge-automower?targetFile=package.json)

[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

<img src="https://user-images.githubusercontent.com/19813688/80208719-1a3dd200-8631-11ea-9e7b-c72abeee8ea9.jpeg" width="25%" align="right"> 
<img src="https://user-images.githubusercontent.com/19813688/80208728-1d38c280-8631-11ea-9169-c5204946db4c.PNG" width="25%" align="right">

Plugin for controlling your [automower](https://www.husqvarna.com/fr/produits/robots-tondeuses/) from [Husqvarna](https://www.husqvarna.com/f) through [HomeBridge](https://github.com/nfarina/homebridge) .

Each Automower is shown through 3 services :

- One switch that will handle Start / PARK (until next period) function
- One fan that will handle Start / Pause function and spin when your mower is cutting (and only cutting, not while getting back to its charging station)
- One Motion sensor that will be triggered when your mower needs attention while cutting

The battery percentage / charging status is shown in the detail pane of each service.

`npm install -g homebridge-automower`

## Homebridge configuration

Config as below:

```json
"platforms": [
  {
    "platform": "Automower",
	  "email": "toto@titi.com",
    "password": "toto"
  }
]
```

Fields:

- `platform` must be "HomebridgeAutomower" (required).
- `email` email used for your automower account (required).
- `password` password of your automower account (required).
- `refreshTimer` Optional - enable refresh of autoMower state every X seconds, for automation purpose if you need to activate something else based on its state change (defaults : disable, accepted range : 30-600s).
- `startStopSwitch` switch to handle start / stop until next period (optional - default true).
- `startStopUntilFurtherNoticeSwitch` stateless switch to handle stop until further notice (optional - default false).
- `mowersList` list of mowers to expose (optional - default all).
- `cleanCache` Set it to true in case you want to remove the cached accessory (only those from this plugin). You have to restart homebridge after applying the option. Remove it after restart, otherwise it will be recreated at each startup.

## Changelog

See [CHANGELOG][].

[changelog]: CHANGELOG.md

## Inspiration

Many thanks to :

- [dinmammas] for plugin inspiration
- [chrisz] for automower api samples
- every tester / contributor that test, and give feedback in any way !

[dinmammas]: https://github.com/dinmammas/homebridge-robonect
[chrisz]: https://github.com/chrisz/pyhusmow

## Donating

Support this project and [others by nicoduj][nicoduj-projects] via [PayPal][paypal-nicoduj].

[![Support via PayPal][paypal-button]][paypal-nicoduj]

[nicoduj-projects]: https://github.com/nicoduj/
[paypal-button]: https://img.shields.io/badge/Donate-PayPal-green.svg
[paypal-nicoduj]: https://www.paypal.me/nicoduj

## License

As of Dec 01 2018, Nicolas Dujardin has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[unlicense]: LICENSE


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 1.4.4

- [FIX] preventing multiple simultaneous auth / refresh

## 1.4.3

- [FIX] warning homebridge 1.3

## 1.4.2

- [FIX] duplicate accessory after restart

## 1.4.1

- [FIX] invalid charging state

## 1.4.0

- [NEW] #21 Add a switch to handle start / stop until further notice
- [FIX] #25 Lifted error not indicated
- [FIX] #24 Parked also means charging
- [NEW] #23 select which mower to use

## 1.3.2

**Platform Name change - you need to uninstall or modify configuration**

- [NEW] cleaning lost devices from cache

## 1.3.1

- [FIX] code improvments

## 1.3.0

**You need to cleanCache if you had installed before, with option cleanCache . Remember to remove it after restart - sorry for that**

- [FIX] better handling of api in order to prevent blocking homekit in case of timeout #16
- [NEW] motion sensor that will trigger in case of error of automower (only if you have background refresh or while refreshing )

## 1.2.2

- [FIX] automatic attemps to discovery if first one fails

## 1.2.1

- [FIX] cleanCache option

## 1.2.0

- [NEW] fix launch error for certification and moving to dynamic platform

## 1.1.9

- [NEW] fix log

## 1.1.8

- [NEW] dep bump and config check for homebridge verified plugin process

## 1.1.7

- [NEW] Supports config UI X configuration interface.

## 1.1.6

- [FIX] fixing #9 Concurrent api logins

## 1.1.5

- [FIX] log fix

## 1.1.4

- [FIX] fixing charging stae / battery level

## 1.1.3

- [NEW] huge refactoring to enhance code quality (I hope there won't be too much bugs ! )
- [NEW] refreshTimer for background refresh

## 1.1.2

- [FIX] charging status not correct

## 1.1.1

- [FIX] status not retrieved

## 1.1.0

- [NEW] #1 Moves to platform mode, so that automower discovery is auto. No need anymore to set the name / model . **CONFIG MUST BE CHANGED**


autoMowerAPI.js
const request = require('request');
//var locks = require('locks');
//var mutex = locks.createMutex();

var EventEmitter = require('events');
const autoMowerConst = require('./autoMowerConst');
var inherits = require('util').inherits;

module.exports = {
  AutoMowerAPI: AutoMowerAPI,
};

function AutoMowerAPI(log, platform) {
  EventEmitter.call(this);

  this.log = log;
  this.platform = platform;
  this.login = platform.login;
  this.password = platform.password;

  this.discoverdMowers = [];

  this.headers = {
    Accept: 'application/json',
    'Content-type': 'application/json',
  };

  this.imApiUrl = 'https://iam-api.dss.husqvarnagroup.net/api/v3/';
  this.trackApiUrl = 'https://amc-api.dss.husqvarnagroup.net/app/v1/';
}

AutoMowerAPI.prototype = {
  logResult: function (result) {
    this.log.debug('INFO - mower status : ' + JSON.stringify(result.status));
    this.log.debug('INFO - mower activity : ' + result.status.mowerStatus.activity);
  },

  authenticate: function (callback) {
    var dte = new Date();

    if (!this.token || (this.token && this.loginExpires && this.loginExpires < dte)) {
      this.log.debug('INFO - authenticating');

      var jsonBody = {
        data: {
          attributes: {
            password: this.password,
            username: this.login,
          },
          type: 'token',
        },
      };

      var that = this;

      //mutex.lock(function () {
      request(
        {
          url: that.imApiUrl + 'token',
          method: 'POST',
          headers: that.headers,
          body: jsonBody,
          json: true,
        },
        function (error, response, body) {
          //mutex.unlock();

          if (error) {
            that.log(error.message);
            callback(error);
          } else if (response && response.statusCode !== 201) {
            that.log(
              'ERROR - authenticate - No 201 return ' +
                response.statusCode +
                '/' +
                JSON.stringify(response)
            );
            callback('No 201');
          } else if (body && body.data) {
            that.token = body.data.id;
            that.tokenProvider = body.data.attributes.provider;
            that.loginExpiry = body.data.attributes.expires_in;
            that.loginExpires = new Date();
            that.loginExpires.setMilliseconds(
              that.loginExpires.getMilliseconds() + that.loginExpiry - 30000
            );
            that.headers['Authorization'] = 'Bearer ' + that.token;
            that.headers['Authorization-Provider'] = that.tokenProvider;
            callback();
          } else {
            that.log('ERROR - authenticate - No body');
            callback('No body');
          }
        }
      );
      //});
    } else {
      this.log.debug('INFO - allready authenticate expiration : ' + this.loginExpires + '-' + dte);
      callback();
    }
  },

  getMowers: function () {
    const that = this;

    if (this.inProgress) return;

    this.inProgress = true;

    this.authenticate((error) => {
      if (error) {
        that.emit('mowersRefreshError');
        that.inProgress = false;
      } else {
        request(
          {
            url: this.trackApiUrl + 'mowers',
            method: 'GET',
            headers: this.headers,
            json: true,
          },
          function (error, response, body) {
            if (error) {
              that.log('ERROR - getMowers - retrieving mower - ' + error.message);
              that.emit('mowersRefreshError');
            } else if (response && response.statusCode !== 200) {
              that.log('ERROR - getMowers - No 200 return ' + response.statusCode + '/' + response);
              that.emit('mowersRefreshError');
            } else if (body.length > 0) {
              let mowers = [];
              body.forEach((mower) => {
                mowers.push(mower);
              });
              that.discoverdMowers = mowers;
              that.emit('mowersUpdated');
            } else {
              that.log('ERROR - getMowers -No body returned from Automower API');
              that.emit('mowersRefreshError');
            }
            that.inProgress = false;
          }
        );
      }
    });
  },

  sendCommand: function (command, callback) {
    const that = this;

    var bodyToSend = {};

    var commandURL = this.trackApiUrl + 'mowers/' + command;

    this.authenticate((error) => {
      if (error) {
        callback(error);
      } else {
        request(
          {
            url: commandURL,
            method: 'POST',
            body: bodyToSend,
            headers: that.headers,
            json: true,
          },
          function (error, response, body) {
            that.log('INFO - Command sent : ' + commandURL);
            that.log.debug('INFO - Body received : ' + JSON.stringify(body));
            if (error) {
              that.log(error.message);
              callback(error);
            } else if (response && response.statusCode !== 200) {
              that.log('ERROR - sendCommand -  No 200 return ' + response.statusCode);
              callback(error);
            } else {
              callback();
            }
          }
        );
      }
    });
  },
};

inherits(AutoMowerAPI, EventEmitter);

//URLS
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/start/';
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/start/';
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/start/override/period'; + JSON String "duration" in body
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/pause';
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/park/duration/timer';
//me.trackApiUrl + 'mowers/' + me.mowerID + '/control/park/duration/period'; + JSON String "duration" in body


index.js
var Service, Characteristic, Accessory, UUIDGen;

var AutoMowerAPI = require('./autoMowerAPI.js').AutoMowerAPI;
const AutoMowerConst = require('./autoMowerConst');

checkTimer = function (timer) {
  if (timer && timer > 0 && (timer < 30 || timer > 600)) return 300;
  else return timer;
};

checkParameter = function (parameter, def) {
  if (parameter == undefined) {
    return def;
  } else {
    if (typeof parameter === 'string') {
      switch (parameter.toLowerCase().trim()) {
        case 'true':
        case 'yes':
          return true;
        case 'false':
        case 'no':
        case null:
          return false;
        case 'undefined':
        default:
          return parameter;
      }
    } else {
      return parameter;
    }
  }
};

function myAutoMowerPlatform(log, config, api) {
  if (!config) {
    log('No configuration found for homebridge-automower');
    return;
  }

  this.api = api;
  this.log = log;
  this.login = config['email'];
  this.password = config['password'];
  this.refreshTimer = checkTimer(config['refreshTimer']);
  this.cleanCache = config['cleanCache'];

  this.startStopSwitch = checkParameter(config['startStopSwitch'], true);
  this.startStopUntilFurtherNoticeSwitch = checkParameter(
    config['startStopUntilFurtherNoticeSwitch'],
    false
  );
  this.mowersList = config['mowersList'];

  this.foundAccessories = [];
  this.autoMowerAPI = new AutoMowerAPI(log, this);

  this.loaded = false;

  this._confirmedAccessories = [];
  this._confirmedServices = [];

  this.api
    .on(
      'shutdown',
      function () {
        this.end();
      }.bind(this)
    )
    .on(
      'didFinishLaunching',
      function () {
        this.log('DidFinishLaunching');

        if (this.cleanCache) {
          this.log('WARNING - Removing Accessories');
          this.api.unregisterPlatformAccessories(
            'homebridge-automower',
            'Automower',
            this.foundAccessories
          );
          this.foundAccessories = [];
        }
        this.discoverAutoMowers();
      }.bind(this)
    );
}

module.exports = function (homebridge) {
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  Accessory = homebridge.platformAccessory;
  UUIDGen = homebridge.hap.uuid;
  HomebridgeAPI = homebridge;
  homebridge.registerPlatform('homebridge-automower', 'Automower', myAutoMowerPlatform, true);
};

myAutoMowerPlatform.prototype = {
  configureAccessory: function (accessory) {
    this.log.debug(accessory.displayName, 'Got cached Accessory ' + accessory.UUID);

    this.foundAccessories.push(accessory);
  },

  end() {
    this.log('INFO - shutdown');
    if (this.timerID) {
      clearInterval(this.timerID);
      this.timerID = undefined;
    }
  },

  //Cleaning methods
  cleanPlatform: function () {
    this.cleanAccessories();
    this.cleanServices();
  },

  cleanAccessories: function () {
    //cleaning accessories
    let accstoRemove = [];
    for (let acc of this.foundAccessories) {
      if (!this._confirmedAccessories.find((x) => x.UUID == acc.UUID)) {
        accstoRemove.push(acc);
        this.log('WARNING - Accessory will be Removed ' + acc.UUID + '/' + acc.displayName);
      }
    }

    if (accstoRemove.length > 0)
      this.api.unregisterPlatformAccessories('homebridge-automower', 'Automower', accstoRemove);
  },

  cleanServices: function () {
    //cleaning services
    for (let acc of this.foundAccessories) {
      let servicestoRemove = [];
      for (let serv of acc.services) {
        if (
          serv.subtype !== undefined &&
          !this._confirmedServices.find((x) => x.UUID == serv.UUID && x.subtype == serv.subtype)
        ) {
          servicestoRemove.push(serv);
        }
      }
      for (let servToDel of servicestoRemove) {
        this.log(
          'WARNING - Service Removed' +
            servToDel.UUID +
            '/' +
            servToDel.subtype +
            '/' +
            servToDel.displayName
        );
        acc.removeService(servToDel);
      }
    }
  },

  discoverAutoMowers: function () {
    this.autoMowerAPI.on('mowersRefreshError', () => {
      if (this.timerID == undefined) {
        this.log('ERROR - discoverAutoMowers - will retry in 1 minute');
        setTimeout(() => {
          this.autoMowerAPI.getMowers();
        }, 60000);
      }
    });

    this.autoMowerAPI.on('mowersUpdated', () => {
      this.log.debug('INFO - mowersUpdated event');
      if (!this.loaded) {
        this.loadMowers();
      } else {
        this.updateMowers();
      }
    });

    this.autoMowerAPI.getMowers();
  },

  loadMowers() {
    let result = this.autoMowerAPI.discoverdMowers;

    if (result && result instanceof Array && result.length > 0) {
      for (let s = 0; s < result.length; s++) {
        let mowerName = result[s].name;

        if (!this.mowersList || (this.mowersList && this.mowersList.includes(mowerName))) {
          let mowerModel = result[s].model + '-' + result[s].variant;
          let mowerSeriaNumber = result[s].id;

          this.log('INFO - Discovered Mower : ' + mowerName);
          this.log.debug('Mower : ' + JSON.stringify(result[s]));

          let uuid = UUIDGen.generate(mowerName);
          let myMowerAccessory = this.foundAccessories.find((x) => x.UUID == uuid);

          if (!myMowerAccessory) {
            myMowerAccessory = new Accessory(mowerName, uuid);
            myMowerAccessory.name = mowerName;
            this.api.registerPlatformAccessories('homebridge-automower', 'Automower', [
              myMowerAccessory,
            ]);
            this.foundAccessories.push(myMowerAccessory);
          }
          myMowerAccessory.model = mowerModel;
          myMowerAccessory.manufacturer = 'Husqvarna Group';
          myMowerAccessory.serialNumber = mowerSeriaNumber;
          myMowerAccessory.mowerID = mowerSeriaNumber;
          myMowerAccessory.name = mowerName;

          myMowerAccessory
            .getService(Service.AccessoryInformation)
            .setCharacteristic(Characteristic.Manufacturer, myMowerAccessory.manufacturer)
            .setCharacteristic(Characteristic.Model, myMowerAccessory.model)
            .setCharacteristic(Characteristic.SerialNumber, myMowerAccessory.serialNumber);

          let HKBatteryService = myMowerAccessory.getServiceByUUIDAndSubType(
            mowerName,
            'BatteryService' + mowerName
          );

          if (!HKBatteryService) {
            this.log('INFO - Creating  Battery Service ' + mowerName);
            HKBatteryService = new Service.BatteryService(mowerName, 'BatteryService' + mowerName);
            HKBatteryService.subtype = 'BatteryService' + mowerName;
            myMowerAccessory.addService(HKBatteryService);
          }

          this.bindBatteryLevelCharacteristic(HKBatteryService);
          this.bindChargingStateCharacteristic(HKBatteryService);
          this.bindStatusLowBatteryCharacteristic(HKBatteryService);

          let HKFanService = myMowerAccessory.getServiceByUUIDAndSubType(
            'Start/Pause ' + mowerName,
            'FanService' + mowerName
          );

          if (!HKFanService) {
            this.log('INFO - Creating  Fan Service ' + mowerName);
            HKFanService = new Service.Fan('Start/Pause ' + mowerName, 'FanService' + mowerName);
            HKFanService.subtype = 'FanService' + mowerName;
            myMowerAccessory.addService(HKFanService);
          }

          this.bindFanOnCharacteristic(myMowerAccessory, HKFanService);

          if (this.startStopSwitch) {
            let HKSwitchService = myMowerAccessory.getServiceByUUIDAndSubType(
              'Start/Park ' + mowerName,
              'SwitchService' + mowerName
            );

            if (!HKSwitchService) {
              this.log('INFO - Creating  Switch Service ' + mowerName);
              HKSwitchService = new Service.Switch(
                'Start/Park ' + mowerName,
                'SwitchService' + mowerName
              );
              HKSwitchService.subtype = 'SwitchService' + mowerName;
              myMowerAccessory.addService(HKSwitchService);
            }

            this.bindSwitchOnCharacteristic(myMowerAccessory, HKSwitchService);
            this._confirmedServices.push(HKSwitchService);
          }

          if (this.startStopUntilFurtherNoticeSwitch) {
            let HKSwitch2Service = myMowerAccessory.getServiceByUUIDAndSubType(
              'ParkUntilResume ' + mowerName,
              'SwitchServicePark' + mowerName
            );

            if (!HKSwitch2Service) {
              this.log('INFO - Creating  Switch Service ' + mowerName);
              HKSwitch2Service = new Service.Switch(
                'ParkUntilResume ' + mowerName,
                'SwitchServicePark' + mowerName
              );
              HKSwitch2Service.subtype = 'SwitchServicePark' + mowerName;
              myMowerAccessory.addService(HKSwitch2Service);
            }

            this.bindSwitchOn2Characteristic(myMowerAccessory, HKSwitch2Service);
            this._confirmedServices.push(HKSwitch2Service);
          }

          let HKMotionService = myMowerAccessory.getServiceByUUIDAndSubType(
            mowerName + ' needs attention',
            'MotionService' + mowerName
          );

          if (!HKMotionService) {
            this.log('INFO - Creating Motion Service ' + mowerName);
            HKMotionService = new Service.MotionSensor(
              mowerName + ' needs attention',
              'MotionService' + mowerName
            );
            HKMotionService.subtype = 'MotionService' + mowerName;
            myMowerAccessory.addService(HKMotionService);
          }
          this.bindMotionCharacteristic(HKMotionService);

          this._confirmedAccessories.push(myMowerAccessory);
          this._confirmedServices.push(HKBatteryService);
          this._confirmedServices.push(HKFanService);

          this._confirmedServices.push(HKMotionService);
        }
      }

      this.cleanPlatform();

      this.updateMowers();
      this.loaded = true;

      //timer for background refresh
      this.refreshBackground();
    } else {
      this.log('ERROR - discoverAutoMowers - no mower found, will retry in 1 minute - ' + result);

      setTimeout(() => {
        this.autoMowerAPI.getMowers();
      }, 60000);
    }
  },

  updateMowers() {
    for (let a = 0; a < this.foundAccessories.length; a++) {
      this.log.debug('INFO - refreshing - ' + this.foundAccessories[a].name);

      let result = this.autoMowerAPI.discoverdMowers;
      let mowerResult = undefined;

      if (result && result instanceof Array && result.length > 0) {
        for (let s = 0; s < result.length; s++) {
          if (result[s].id === this.foundAccessories[a].mowerID) {
            mowerResult = result[s];
            break;
          }
        }
      }

      if (mowerResult !== undefined) {
        this.refreshAutoMower(this.foundAccessories[a], mowerResult);
      } else {
        this.log('ERROR - updateMowers - no result for mower - ' + this.foundAccessories[a].name);
      }
    }
  },

  getBatteryLevel(homebridgeAccessory, result) {
    var percent = 0;

    if (result && result.status) {
      percent = result.status.batteryPercent;
    }

    return percent;
  },

  getBatteryLevelCharacteristic: function (service, callback) {
    this.log.debug('INFO - getBatteryLevelCharacteristic');

    var percent = service.getCharacteristic(Characteristic.BatteryLevel).value;
    callback(undefined, percent);

    this.autoMowerAPI.getMowers();
  },

  getChargingState(homebridgeAccessory, result) {
    var charging = 0;
    if (
      result &&
      result.status &&
      result.status.connected &&
      result.status.batteryPercent < 100 &&
      (result.status.mowerStatus.activity.startsWith(AutoMowerConst.CHARGING) ||
        result.status.mowerStatus.activity.startsWith(AutoMowerConst.PARKED))
    ) {
      charging = 1;
    }
    this.log.debug('INFO - getChargingState -' + charging);
    return charging;
  },

  getChargingStateCharacteristic: function (service, callback) {
    this.log.debug('INFO - getChargingStateCharacteristic');
    var charging = 0;

    var charging = service.getCharacteristic(Characteristic.ChargingState).value;
    callback(undefined, charging);

    //no update asked, handled by batteryLevel
  },

  isLowBattery(homebridgeAccessory, result) {
    var lowww = 0;

    if (result && result.status && result.status.batteryPercent < 20) {
      lowww = 1;
    }

    return lowww;
  },

  getLowBatteryCharacteristic: function (service, callback) {
    this.log.debug('INFO - getLowBatteryCharacteristic');
    var lowww = 0;

    var lowww = service.getCharacteristic(Characteristic.StatusLowBattery).value;
    callback(undefined, lowww);

    //no update asked, handled by batteryLevel
  },

  isInOperation(homebridgeAccessory, result) {
    var onn = false;

    if (
      result &&
      result.status &&
      result.status.mowerStatus.state.startsWith(AutoMowerConst.IN_OPERATION)
    ) {
      onn = true;
    }

    return onn;
  },

  getSwitchOnCharacteristic: function (service, callback) {
    this.log.debug('INFO - getSwitchOnCharacteristic');
    var onn = false;

    var onn = service.getCharacteristic(Characteristic.On).value;
    callback(undefined, onn);

    this.autoMowerAPI.getMowers();
  },
  setSwitchOnCharacteristic: function (homebridgeAccessory, characteristic, value, callback) {
    this.log.debug('INFO - setSwitchOnCharacteristic - ' + value);

    var currentValue = characteristic.value;
    callback();

    this.autoMowerAPI.sendCommand(
      homebridgeAccessory.mowerID +
        (value ? AutoMowerConst.START_COMMAND : AutoMowerConst.PARK_COMMAND),
      function (error) {
        if (error) {
          setTimeout(function () {
            characteristic.updateValue(currentValue);
          }, 200);
        }
      }
    );
  },

  setSwitchOn2Characteristic: function (homebridgeAccessory, characteristic, value, callback) {
    this.log.debug('INFO - setSwitchOn2Characteristic - ' + value);

    var currentValue = characteristic.value;
    callback();

    this.autoMowerAPI.sendCommand(
      homebridgeAccessory.mowerID + AutoMowerConst.PARK_COMMAND_UFN,
      function (error) {
        setTimeout(function () {
          characteristic.updateValue(false);
        }, 200);
      }
    );
  },

  isMowing(homebridgeAccessory, result) {
    var mowing = 0;

    if (result && result.status) {
      let status = result.status.mowerStatus.activity;
      if (status.startsWith(AutoMowerConst.MOWING)) mowing = 1;
      else if (status.startsWith(AutoMowerConst.STOPPED_IN_GARDEN)) mowing = -1;
      else if (
        result.status.mowerStatus.state &&
        result.status.mowerStatus.state.startsWith(AutoMowerConst.FATAL_ERROR)
      )
        mowing = -1;
    }

    return mowing;
  },

  getMowerOnCharacteristic: function (service, callback) {
    this.log.debug('getMowerOnCharacteristic');

    var mowing = service.getCharacteristic(Characteristic.On).value;
    callback(undefined, mowing);

    this.autoMowerAPI.getMowers();
  },
  setMowerOnCharacteristic: function (homebridgeAccessory, characteristic, value, callback) {
    this.log.debug('setMowerOnCharacteristic -' + value);

    var currentValue = characteristic.value;

    callback();

    this.autoMowerAPI.sendCommand(
      homebridgeAccessory.mowerID +
        (value ? AutoMowerConst.START_COMMAND : AutoMowerConst.PAUSE_COMMAND),
      function (error) {
        if (error) {
          setTimeout(function () {
            characteristic.updateValue(currentValue);
          }, 200);
        }
      }
    );
  },

  bindBatteryLevelCharacteristic: function (service) {
    service.getCharacteristic(Characteristic.BatteryLevel).on(
      'get',
      function (callback) {
        this.getBatteryLevelCharacteristic(service, callback);
      }.bind(this)
    );
  },

  bindChargingStateCharacteristic: function (service) {
    service.getCharacteristic(Characteristic.ChargingState).on(
      'get',
      function (callback) {
        this.getChargingStateCharacteristic(service, callback);
      }.bind(this)
    );
  },

  bindStatusLowBatteryCharacteristic: function (service) {
    service.getCharacteristic(Characteristic.StatusLowBattery).on(
      'get',
      function (callback) {
        this.getLowBatteryCharacteristic(service, callback);
      }.bind(this)
    );
  },

  bindFanOnCharacteristic: function (homebridgeAccessory, service) {
    service
      .getCharacteristic(Characteristic.On)
      .on(
        'get',
        function (callback) {
          this.getMowerOnCharacteristic(service, callback);
        }.bind(this)
      )
      .on(
        'set',
        function (value, callback) {
          this.setMowerOnCharacteristic(
            homebridgeAccessory,
            service.getCharacteristic(Characteristic.On),
            value,
            callback
          );
        }.bind(this)
      );
  },

  bindSwitchOnCharacteristic: function (homebridgeAccessory, service) {
    service
      .getCharacteristic(Characteristic.On)
      .on(
        'get',
        function (callback) {
          this.getSwitchOnCharacteristic(service, callback);
        }.bind(this)
      )
      .on(
        'set',
        function (value, callback) {
          this.setSwitchOnCharacteristic(
            homebridgeAccessory,
            service.getCharacteristic(Characteristic.On),
            value,
            callback
          );
        }.bind(this)
      );
  },

  bindSwitchOn2Characteristic: function (homebridgeAccessory, service) {
    service
      .getCharacteristic(Characteristic.On)
      .on(
        'get',
        function (callback) {
          callback(undefined, false);
        }.bind(this)
      )
      .on(
        'set',
        function (value, callback) {
          this.setSwitchOn2Characteristic(
            homebridgeAccessory,
            service.getCharacteristic(Characteristic.On),
            value,
            callback
          );
        }.bind(this)
      );
  },

  bindMotionCharacteristic(service) {
    service.getCharacteristic(Characteristic.MotionDetected).on(
      'get',
      function (callback) {
        callback(undefined, false);
      }.bind(this)
    );
  },

  refreshBackground() {
    //timer for background refresh
    if (this.refreshTimer !== undefined && this.refreshTimer > 0) {
      this.log.debug(
        'INFO - Setting Timer for background refresh every  : ' + this.refreshTimer + 's'
      );
      this.timerID = setInterval(() => this.autoMowerAPI.getMowers(), this.refreshTimer * 1000);
    }
  },

  refreshAutoMower: function (myAutoMowerAccessory, result) {
    let mowerName = myAutoMowerAccessory.name;

    this.autoMowerAPI.logResult(result);

    let HKSwitchService = myAutoMowerAccessory.getServiceByUUIDAndSubType(
      'Start/Park ' + mowerName,
      'SwitchService' + mowerName
    );

    if (HKSwitchService) {
      HKSwitchService.getCharacteristic(Characteristic.On).updateValue(
        this.isInOperation(myAutoMowerAccessory, result)
      );
    }

    let HKSwitchService2 = myAutoMowerAccessory.getServiceByUUIDAndSubType(
      'Start/ParkUntilResume ' + mowerName,
      'SwitchService2' + mowerName
    );

    if (HKSwitchService2) {
      HKSwitchService2.getCharacteristic(Characteristic.On).updateValue(
        this.isInOperation(myAutoMowerAccessory, result)
      );
    }

    let HKFanService = myAutoMowerAccessory.getServiceByUUIDAndSubType(
      'Start/Pause ' + mowerName,
      'FanService' + mowerName
    );

    if (HKFanService) {
      let isMowing = this.isMowing(myAutoMowerAccessory, result);
      HKFanService.getCharacteristic(Characteristic.On).updateValue(isMowing > 0 ? true : false);

      let HKMotionService = myAutoMowerAccessory.getServiceByUUIDAndSubType(
        mowerName + ' needs attention',
        'MotionService' + mowerName
      );

      //handling error
      if (HKMotionService) {
        if (isMowing < 0) {
          HKMotionService.getCharacteristic(Characteristic.MotionDetected).updateValue(true);
          this.log.debug('INFO - MotionDetected');
        } else if (HKMotionService.getCharacteristic(Characteristic.MotionDetected).value) {
          HKMotionService.getCharacteristic(Characteristic.MotionDetected).updateValue(false);
          this.log.debug('INFO - MotionDetected reset ');
        }
      }
    }

    let HKBatteryService = myAutoMowerAccessory.getServiceByUUIDAndSubType(
      mowerName,
      'BatteryService' + mowerName
    );

    if (HKBatteryService) {
      HKBatteryService.getCharacteristic(Characteristic.BatteryLevel).updateValue(
        this.getBatteryLevel(myAutoMowerAccessory, result)
      );
      HKBatteryService.getCharacteristic(Characteristic.ChargingState).updateValue(
        this.getChargingState(myAutoMowerAccessory, result)
      );
      HKBatteryService.getCharacteristic(Characteristic.StatusLowBattery).updateValue(
        this.isLowBattery(myAutoMowerAccessory, result)
      );
    }
  },
};
